### PR TITLE
Add Uniswapv3 liquidity to bad token detector

### PR DESCRIPTION
**Pull request by vkgnosis**
There have been several tokens that we detect as bad because the bad token detector needs an account with funding and couldn't find one on uniswapv2 and balancer. This PR improves the situation by allowing it to consider uniswapv3 pools.

### Test Plan
CI, new unit test, new ignored bad token test

### CODE REVIEW

**Review by nlordell — COMMENTED** [review]
Very cool and hopefully gets us a few less false positives on "bad tokens".

Would like to see possible fee values being fetched instead of hardcoded, but if it adds too much work then I'm also fine with punting it until its needed.

File: crates/shared/src/bad_token/trace_call.rs
Comment (vkgnosis) [reply crates/shared/src/bad_token/trace_call.rs:64]: Will do in another PR.

**Review by nlordell — APPROVED** [review]

File: crates/contracts/src/bin/vendor.rs
Comment (nlordell) [reply crates/contracts/src/bin/vendor.rs:129]: My guess is that L127 is too long which causes issues with `rustfmt`. This is why, for some of the Balancer contracts I used `\` to break the string into multiple lines.

**Comment by bh2smith** [comment]
Do we have a particular token in mind that brought this to the surface more recently? I would specifically like to know if these changes resolve the issue of false bad token detecting.

**Comment by fleupold** [comment]
I'm worried that this change as is will make the user experience worse as it may lead the system to believe tokens are tradable (when they are not) and lead to users placing orders which eventually are bound to expire unmatched.

In particular, if there is a token for which only a uni v3 pool exists they are currently marked as unsupported and users cannot place an order to begin with (which is a bad experience but not as frustrating as placing an order that doesn't get matched). 
After this change, users would be able to place an order but their order would not be matchable because we don't have a baseline price estimate for univ3 only tokens (technically they can still get matched if the orderbook is completely empty but that's a rare case in prod). 

This doesn't make this change wrong, but I think we should consider if we want to prevent these types of orders altogether (until we pass down aggregator based API price estimations with the `solvable_orders` endpoint). That being said, it's probably cleaner to explicitly block these orders elsewhere than relying on the bad_token detector (but the two PRs might have to go together).

File: crates/orderbook/src/main.rs
Comment (MartinquaXD) [inline crates/orderbook/src/main.rs:348]: nit: This looks unusual. Since you are only handling `NotFound` and no other error, I assume `other => Some(other.unwrap())` is supposed to throw on every error except `NotFound`.
Could this be made more explicit with:
```rust
match IUniswapV3Factory::deployed(&web3).await {
    Ok(factory) => finders.push(Arc::new(UniswapV3Finder {
        factory,
        base_tokens: base_tokens.tokens().iter().copied().collect(),
    })),
    Err(DeployError::NotFound(_)) => (),
    Err(e) => panic!("{}", e.to_string()),
};
```

**Review by MartinquaXD — APPROVED** [review]
Changes look good. I like that you included a link to the solidity code computing pool addresses. Otherwise I would have been a little lost. 👍

**Comment by vkgnosis** [comment]
> At the moment, none of the tests introduced here give me any indication that this does indeed succeed in correctly detecting a tradable token for which there is only uniV3 liquidity.

I'll add a token that has only v3 liquidity to the test. I'm being pedantic here but I can't follow this line of reasoning. The test shows that univ3 pools are considered by the detector because it still detects an example token as good when using only univ3 liquidity. Whether there is or isn't univ2 liquidity is not relevant for showing that the univ3 functionality works.

> I'm worried that this change as is will make the user experience worse as it may lead the system to believe tokens are tradable (when they are not) and lead to users placing orders which eventually are bound to expire unmatched.

Feels like we've discussed this before but I didn't remember when I made the PR. If we want to say that we do not accept orders for tokens that don't have univ2 pools then we should make that explicit somewhere. It's not the job of the bad token detector. We already have this problem because we consider balancer for the detector. Since we seem to agree on that then I'd say before merging this we should either create that component or improve the situation in the driver so that it doesn't need baseline estimates.

**Comment by nlordell** [comment]
> It's not the job of the bad token detector

I also agree with this.

> Since we seem to agree on that then I'd say before merging this we should either create that component or improve the situation in the driver so that it doesn't need baseline estimates.

The issue we have now is that the `orderbook` doesn't consider Balancer V2 liquidity in the `Baseline` solver. This would mean that, currently, it will be stricter than it needs to. 

Also, it may be worth talking about the points you brought up [here](https://github.com/gnosis/gp-v2-services/issues/1571#issuecomment-1014338082) where we should discuss if we want to continue with `Baseline` solver for external price estimates or move to `Quasimodo`. I'm in favour of the latter.

Finally, I wanted to link an issue for improving price augmentation logic in the solver: https://github.com/gnosis/gp-v2-services/issues/1537.

**Comment by nlordell** [comment]
Just posting an update here.

Since we decided to move towards using price estimators for computing external prices (see #1571) we no longer have to worry about accepting orders that we don't support.

I think we should hold off on merging until we have the external prices per auction computed in the `orderbook` process.

**Comment by nlordell** [comment]
We can merge this once #1649 is merged I think.

**Comment by vkgnosis** [comment]
Your previous comment was probably referring to https://github.com/gnosis/gp-v2-services/pull/1655 .
That has been merged so feel free to merge this whenever.

**Pull request merged**
